File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cs
public class Solution {
    public int NumberOfSubstrings(string s) {
        int n = s.Length;
        int[] dict = new int[3];
        
        bool check() {
            return dict[0] > 0 && dict[1] > 0 && dict[2] > 0;
        }
        
        int ans = 0;
        
        int left = 0;
        for (int i = 0; i < n; i++) {
            dict[s[i] - 'a']++;
            while (check()) { 
                ans += n - i;
                dict[s[left++] - 'a']--;
            }
        }
        
        return ans;
    }
}

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        dict = {'a': 0, 'b': 0, 'c': 0}
        
        ans = 0
        
        left = 0
        for i in range(n):
            dict[s[i]] += 1
            
            while dict['a'] >= 1 and dict['b'] >= 1 and dict['c'] >= 1:
                ans += n - i
                dict[s[left]] -= 1
                left += 1
        
        return ans